### DISCUSSION TITLE
AUPRC에 관한 고찰...

### DISCUSSION THREAD

**Discussion by mountinyy** [boostcampaitech4lv23nlp2/level2_klue_nlp-level2-nlp-10#15] [General]
Precision, Recall, AUPRC에 대해서 이미 아시면 바로 문제점 부분을 봐주시면 될 것 같습니다!
# Precision, Recall
![image](https://user-images.githubusercontent.com/28773464/203685337-7a19511d-f665-478c-9c8c-0d6e864ff7c1.png)
- TP : 모델이 Positive라고 예측해서 맞춘(True)개수 (실제 정답은 Positive)
- TN : 모델이 Negative라고 예측해서 맞춘(False) 개수 (실제 정답은 Negative)
- FP : 모델이 Positive라고 예측해서 틀린(False)개수 (실제 정답은 Negative)
- FN : 모델이 Negative라고 예측해서 틀린(False)개수 (실제 정답은 Positive)

이를 사용해 Precision, Recall을 구하는 식은 다음과 같습니다.
![image](https://user-images.githubusercontent.com/28773464/203685652-99f9d852-ef71-408d-8bd5-1849bdc9f2bd.png)
즉, 모델이 Postiive라고 예측한 결과 중 실제로 맞춘 비율입니다.

![image](https://user-images.githubusercontent.com/28773464/203685665-b780e2a6-297b-4dcb-b40a-6276ef6e8ac3.png)
즉, 실제 정답이 Positive인 것들 중 모델이 실제로 맞춘 비율입니다.

(Precision, Recall은 이진 분류에서만 사용 가능하기 때문에 저희 코드에서는 정답 label을 one-hot 벡터를 사용하여 표현합니다.)
# Threhsold 적용
우리가 다루고 있는 task처럼 Positive, Negative를 확률로 구분할 경우 threshold라는 개념이 사용됩니다.
앞서 얘기한 Precision, Recall은 이 threshold에 따라 값이 바뀌게 되며, 여러 threshold를 사용하여 도출한 여러 (Precision, Recall) 쌍으로 그래프를 그린 것이 저희가 구하는 precision_recall_curve입니다.
# Threhsold에 따른 Precision, Recall 계산
다음과 같이 정답 Label과 우리 모델이 예측한 Probability가 있다고 가정하겠습니다.
```python3
label = [0, 0, 1, 1],
probability = [0.1, 0.4, 0.35, 0.8]
```
### 1. 우선 (label, probability) 쌍을 probability가 높은 순으로 재정렬합니다.
```python3
label = [1, 0, 1, 0],
probability = [0.8, 0.4, 0.35, 0.1]
```
### 2. 가장 높은 probability를 threshold로 설정합니다. (threshold=0.8)
그리고 probability >= threshold 인 경우 positive로 예측하고, 아니면 negative로 예측합니다.
<html>
<body>

Label | Probability | Prediction
-- | -- | --
positive | 0.80 | positive
negative | 0.40 | negative
positive | 0.35 | negative
negative | 0.10 | negative

</body>
</html>

이 경우 TP=1, FP=0, FN=1이므로
- precision = 1 / (1+0) = 1
- recall = 1 / (1+1) = 0.5

∴ threshold=0.8일 때의 (recall, precision)은 (0.5, 1)이 됩니다.
### 3. 다음으로 가장 높은 probability를 threshold로 설정합니다. (threshold=0.4)
2번과 마찬가지로 probability >= threshold인 경우 positive로 예측하고, 아니면 negative로 예측합니다.
<html>
<body>

Label | Probability | Prediction
-- | -- | --
positive | 0.80 | positive
negative | 0.40 | positive
positive | 0.35 | negative
negative | 0.10 | negative

</body>
</html>

이 경우 TP=1, FP=1, FN=1이므로 
- precision = 1 / (1+1) = 0.5
- recall = 1 / (1+1) = 0.5

∴ threshold=0.4일 때의 (recall, precision)은  (0.5, 0.5)이 됩니다.
### 4. 가장 낮은 probability를 threshold로 설정할 때까지 2번 3번을 반복하여 (recall, precision)을 구합니다.
- (0.5, 1), (0.5, 0.5), (1, 0.66667), (1, 0.5)
### 5. (recall, precision)들을 이용하여 precision_recall_curve를 그립니다. (x축 = recall, y축 = precision)
![download](https://user-images.githubusercontent.com/28773464/203691397-0980f0a3-dee4-499f-82fe-9c850660469d.png)
그림이 조금 이상하지만..정상적인 데이터라면 아래 그림과 비슷하겠죠..?
![image](https://user-images.githubusercontent.com/28773464/203691566-fe9ac0c9-6861-44f3-b123-146cd32c00ae.png)

여기서 그래프 밑의 넓이가 우리가 구하고자 하는 AUPRC(Area Under Precision-Recall Curve)입니다!
다만 저희 task의 경우 이진 분류가 아닌 다중 분류이기 때문에 30개의 클래스에 대한 각각의 AUPRC를 구하고, 이 값들의 평균을 최종 AUPRC로 사용합니다.

# 문제점
저희 문제는 recall값이 nan이 나타난다는 것이죠.
그 이유는 label 값이 모두 0(Negative)일 경우 recall을 계산할 때 0으로 나뉘어지기 때문입니다.
<html>
<body>

Label | Probability | Prediction
-- | -- | --
negative | 0.80 | positive
negative | 0.40 | positive
negative | 0.35 | negative
negative | 0.10 | negative

</body>
</html>

위와 같이 label이 모두 negative인 상황에서는 threshold 값에 관계없이 무조건 TP=0, FN=0이 계산되기 때문에 
- recall = 0 / (0+0)

으로 계산되어 nan이 나타납니다.
이후 nan값을 사용한 사칙연산은 무조건 계산결과가 nan값이 나오게 됩니다.

auprc를 구하는 저희 baseline 코드를 보시면 다음과 같습니다.
```python3
def klue_re_auprc(probs, labels): # probs.shape : (batch_size, class개수), labels.shape : (batch_size)
    """KLUE-RE AUPRC (with no_relation)"""
    labels = np.eye(30)[labels] # label을 one-hot 인코딩 형식으로 바꿔줌. (shape : (batch_size, class개수))
    score = np.zeros((30,))
    
    for c in range(30):
        # .take().ravel() : labels에서 c 클래스에 대한 값을 모든 배치에서 가져옴(shape : [batch_size, 1]) -> ravel()(shape : [batch_size,])
        targets_c = labels.take([c], axis=1).ravel() 
        preds_c = probs.take([c], axis=1).ravel()
        precision, recall, _ = sklearn.metrics.precision_recall_curve(targets_c, preds_c)
        score[c] = sklearn.metrics.auc(recall, precision)
    return np.average(score) * 100.0
```
위 코드에서 for문을 보면 30개의 클래스에 대해 loop를 도는데, targets_c를 아래 그림과 같이 가져옵니다. (batch_size=5)

![image](https://user-images.githubusercontent.com/28773464/203695675-0d921737-c247-478b-aef9-e045f753421b.png)

preds_c도 위와 비슷한 방식으로 가져오되, 값들은 {0, 1}이 아닌 [0, 1]이라는 점만 다릅니다.
그러나 만약 c=2인 경우에는 

![image](https://user-images.githubusercontent.com/28773464/203695955-5ef20ec6-1798-4524-9be4-c6b61057c3d1.png)

위 그림처럼 targets_c가 설정되어 모든 label값이 0이 되고, 이 경우 recall값이 nan으로 계산됩니다.

# 논의사항
- nan값을 어떻게 처리하면 좋을까요..?
